libr/bin/dwarf.c

```diff
@@ -190,6 +190,72 @@ static const char *dwarf_langs[] = {
 	[DW_LANG_Fortran08] = "Fortran08"
 };
 
+typedef struct {
+	const ut8 *buf;
+	st64 len;
+	bool big_endian;
+	size_t address_size;
+} data_xtor;
+
+void data_xtor_init(data_xtor *e, const ut8 *buf, st64 len, bool big_endian) {
+	e->buf = buf;
+	e->len = len;
+	e->big_endian = big_endian;
+}
+
+ut8 data_xtor_u8(data_xtor *e, st64 *off) {
+	ut8 ret = 0;
+	if (*off < e->len) {
+		ret = e->buf[*off];
+		*off += 1;
+	}
+	return ret;
+}
+
+ut16 data_xtor_u16(data_xtor *e, st64 *off) {
+	ut16 ret = 0;
+	if (*off + 2 <= e->len) {
+		ret = *(uut16*)(e->buf + *off);
+		if (R_SYS_ENDIAN != (int)e->big_endian)
+			ret = r_swap_ut16 (ret);
+		*off += 2;
+	}
+	return ret;
+}
+
+ut32 data_xtor_u32(data_xtor *e, st64 *off) {
+	ut32 ret = 0;
+	if (*off + 4 <= e->len) {
+		ret = *(uut32*)(e->buf + *off);
```

Comment:
use r_read_ apis instead of checking, swaping comparing

Reply:
No. This is uncommon unaligned read which may not be necessary to be a public API in r_endian.h

Reply:
r_read apis don't care about alignments, is this because of the type punned thing because its trying to read beyond the end of the buffer?

Reply:
because it looks like an overengineered fix to me